libs/dbus_util/inc/public/tfc/dbus/sml_interface.hpp

```diff
@@ -0,0 +1,219 @@
+    return detail::extract_event_type(event);
+  } else {
+    return boost::sml::aux::get_type_name<event_t>();
+  }
+}
+
+}  // namespace detail
+
+/// \brief Interface for state machine logging and dbus API
+/// \example
+/// \code{.cpp}
+/// #include <tfc/dbus/sml_interface.hpp>
+/// struct state_machine {
+///   auto operator()() {
+///     using boost::sml::operator""_s;
+///     using boost::sml::operator""_e;
+///     return boost::sml::make_transition_table(
+///         * "init"_s + "set_stopped"_e = "stopped"_s
+///     );
+///   }
+/// };
+///
+/// int main() {
+///   boost::asio::io_context ctx{};
+///   auto bus = std::make_shared<sdbusplus::asio::connection>(ctx);
+///   auto interface = std::make_shared<sdbusplus::asio::dbus_interface>(bus, std::string{ tfc::dbus::sml::tags::path },
+///   "StateMachineName"); tfc::dbus::sml::interface sml_interface{ interface, "Log key" }; // optional log key
+///   // NOTE! interface struct requires to be passed by l-value like below, so the using code needs to store it like above
+///   boost::sml::sm<state_machine, boost::sml::logger<tfc::dbus::sml::interface>> sm{ sml_interface };
+///   return EXIT_SUCCESS;
+/// }
+/// \endcode
+/// Get from cli example:
+/// busctl --system get-property com.skaginn3x.tfc.operations.def /com/skaginn3x/StateMachines com.skaginn3x.Operations
+/// StateMachine
+struct interface : tfc::logger::sml_logger {
+  using logger = tfc::logger::sml_logger;
+
+  explicit interface(std::shared_ptr<sdbusplus::asio::dbus_interface> interface) : logger{}, impl_{ std::move(interface) } {}
+  explicit interface(std::shared_ptr<sdbusplus::asio::dbus_interface> interface, std::string_view log_key)
+      : logger{ log_key }, impl_{ std::move(interface) } {}
+  interface(interface const&) = delete;
+  interface(interface&&) noexcept = default;
+  auto operator=(interface const&) -> interface& = delete;
+  auto operator=(interface&&) noexcept -> interface& = default;
+
+  template <class state_machine_t, class event_t>
+  void log_process_event([[maybe_unused]] event_t const& event) {
+    last_event_ = detail::extract_event_type(event);
+    logger::log_process_event<state_machine_t>(event);
+  }
+
+  template <class state_machine_t, class guard_t, class event_t>
+  void log_guard(guard_t const& guard, event_t const& event, bool result) {
+    logger::log_guard<state_machine_t>(guard, event, result);
+  }
+
+  template <class state_machine_t, class action_t, class event_t>
+  void log_action(action_t const& action, event_t const& event) {
+    logger::log_action<state_machine_t>(action, event);
+  }
+
+  template <class state_machine_t, class source_state_t, class destination_state_t>
+  void log_state_change(source_state_t const& src, destination_state_t const& dst) {
+    impl_.on_state_change(src.c_str(), dst.c_str(), last_event_);
+    std::stringstream iss{};
+    detail::dump<boost::sml::sm<state_machine_t>>(dst, iss);
+    impl_.dot_format(iss.str());
```

Comment:
Can we do this without a stringstream?

Reply:
yes, I just copied the example from boost sml which used stringstream, it should actually be quicker to use string https://stackoverflow.com/questions/19844858/c-stdostringstream-vs-stdstringappend so I will change to string

Reply:
fixed here https://github.com/Skaginn3x/framework/pull/225/commits/b1b60c17de98847cd8073daa9b5a3abae14ad128